src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1233,137 @@ private int SkipBlockComment(int i)
+            }
+
+            int maxNumberOfHexDigits = (isBracketedSequence ? 6 : 4);
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            for (int i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = GetChar();
+
+                // Bracketed sequence has been terminated.
+                if (isBracketedSequence && (c == '}'))
+                {
+                    if (i == 0)
+                    {
+                        // Variable sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidBracketedUnicodeEscapeSequence);
+                        return s_unknownChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if (!c.IsHexDigit())
+                {
+                    UngetChar();
+
+                    if (isBracketedSequence)
+                    {
+                        ReportError(_currentIndex, () => ParserStrings.UnicodeEscapeSequenceMissingTerminator);
+                    }
+                    else
+                    {
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                    }
+
+                    return s_unknownChar;
+                }
+
+                sb.Append(c);
+            }
+
+            if (isBracketedSequence && !isTerminated)
+            {
+                c = GetChar();
+                if (c != '}')
+                {
+                    UngetChar();
+                    ReportError(_currentIndex, () => ParserStrings.UnicodeEscapeSequenceMissingTerminator);
+                    return s_unknownChar;
+                }
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue = int.Parse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo);
+            if (unicodeValue <= Char.MaxValue)
+            {
+                return ((char)unicodeValue).ToString();
+            }
+            else if (unicodeValue <= 0x10FFFF)
+            {
+                return Char.ConvertFromUtf32(unicodeValue);
+            }
+            else
+            {
+                // Place the error indicator under only the hex digits in the esc sequence.
+                IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex + 3, _currentIndex - 1);
+                ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequenceValue);
+                return s_unknownChar;
             }
         }
 
+        private bool IsSurrogatePair(string str, ref char nonSurrogatePairChar, StringBuilder sb1 = null, StringBuilder sb2 = null) 
```

Comment:
Just a thought - is this cleaner if you use overloads instead of default parameters? It would avoid the null tests.